mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/quota/CassandraCurrentQuotaManagerV2.java

```diff
@@ -0,0 +1,129 @@
+    }
+
+    @Override
+    public Mono<QuotaCountUsage> getCurrentMessageCount(QuotaRoot quotaRoot) {
+        return currentValueDao.getQuotaCurrentValue(asQuotaKeyCount(quotaRoot))
+            .map(QuotaCurrentValue::getCurrentValue)
+            .map(QuotaCountUsage::count)
+            .defaultIfEmpty(QuotaCountUsage.count(0L));
+    }
+
+    @Override
+    public Mono<QuotaSizeUsage> getCurrentStorage(QuotaRoot quotaRoot) {
+        return currentValueDao.getQuotaCurrentValue(asQuotaKeySize(quotaRoot))
+            .map(QuotaCurrentValue::getCurrentValue)
+            .map(QuotaSizeUsage::size)
+            .defaultIfEmpty(QuotaSizeUsage.size(0L));
+    }
+
+    @Override
+    public Mono<CurrentQuotas> getCurrentQuotas(QuotaRoot quotaRoot) {
+        return currentValueDao.getQuotasByComponent(QuotaComponent.MAILBOX, quotaRoot.asString())
+            .collectList()
+            .map(this::buildCurrentQuotas);
+    }
+
+    private CurrentQuotas buildCurrentQuotas(List<QuotaCurrentValue> quotaCurrentValues) {
+        QuotaCountUsage count = extractQuotaByType(quotaCurrentValues, QuotaType.COUNT)
+            .map(value -> QuotaCountUsage.count(value.getCurrentValue()))
+            .orElse(QuotaCountUsage.count(0L));
+
+        QuotaSizeUsage size = extractQuotaByType(quotaCurrentValues, QuotaType.SIZE)
+            .map(value -> QuotaSizeUsage.size(value.getCurrentValue()))
+            .orElse(QuotaSizeUsage.size(0L));
+
+        return new CurrentQuotas(count, size);
+    }
+
+    private Optional<QuotaCurrentValue> extractQuotaByType(List<QuotaCurrentValue> quotaCurrentValues, QuotaType quotaType) {
+        return quotaCurrentValues.stream()
+            .filter(quotaValue -> quotaValue.getQuotaType().equals(quotaType))
+            .findAny();
+    }
+
+    @Override
+    public Mono<Void> increase(QuotaOperation quotaOperation) {
+        return currentValueDao.increase(asQuotaKeyCount(quotaOperation.quotaRoot()), quotaOperation.count().asLong())
+            .then(currentValueDao.increase(asQuotaKeySize(quotaOperation.quotaRoot()), quotaOperation.size().asLong()));
+    }
+
+    @Override
+    public Mono<Void> decrease(QuotaOperation quotaOperation) {
+        return decrease(quotaOperation.quotaRoot(), quotaOperation.count(), quotaOperation.size());
+    }
+
+    private Mono<Void> decrease(QuotaRoot quotaRoot, QuotaCountUsage count, QuotaSizeUsage size) {
+        return currentValueDao.decrease(asQuotaKeyCount(quotaRoot), count.asLong())
+            .then(currentValueDao.decrease(asQuotaKeySize(quotaRoot), size.asLong()));
+    }
+
+    @Override
+    public Mono<Void> setCurrentQuotas(QuotaOperation quotaOperation) {
+        return getCurrentQuotas(quotaOperation.quotaRoot())
+            .filter(Predicate.not(Predicate.isEqual(CurrentQuotas.from(quotaOperation))))
+            .flatMap(storedQuotas -> decrease(quotaOperation.quotaRoot(), storedQuotas.count(), storedQuotas.size())
```

Comment:
`decrease` then `increase` => we need 4 queries
I think we can evaluate the "changed" before query to the database 
Ex: 
```suggestion
             .flatMap(storedQuotas -> {
                long count = quotaOperation.count().asLong() - storedQuotas.count().asLong();
                long size = quotaOperation.size().asLong() - storedQuotas.size().asLong();

                return currentValueDao.increase(asQuotaKeyCount(quotaOperation.quotaRoot()), count)
                    .then(currentValueDao.increase(asQuotaKeySize(quotaOperation.quotaRoot()), size));
            });
```
we need 2 queries with this way

Reply:
cf fixed in https://github.com/apache/james-project/pull/1702/commits/85bef866fb552c5935b0273782f232dae61e2806

Didnt want to make a fixup, the CI seems too capricious lately, would be happy to not have to squash and wait again after a green build :)